sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala

```diff
@@ -331,20 +346,51 @@ class RocksDBFileManager(
     }
   }
 
+  /**
+   * Based on the ground truth lineage loaded from changelog file (lineage), this function
+   * does file listing to find all snapshot (version, uniqueId) pairs, and finds
+   * the ground truth latest snapshot (version, uniqueId) the db instance needs to load.
+   *
+   * @param lineage the ground truth lineage loaded from changelog file
+   * @return the ground truth latest snapshot (version, uniqueId) the db instance needs to load
+   */
+  def getLatestSnapshotVersionAndUniqueIdFromLineage(
+      lineage: Array[LineageItem]): Option[(Long, Option[String])] = {
+    val path = new Path(dfsRootDir)
+    if (fm.exists(path)) {
+      fm.list(path, onlyZipFiles)
+        .map(_.getPath.getName.stripSuffix(".zip").split("_"))
+        .filter {
+          case Array(ver, id) => lineage.contains(LineageItem(ver.toLong, id))
+        }
+        .map {
+          case Array(version, uniqueId) => (version.toLong, Option(uniqueId))
+        }
+        .sortBy(_._1)
+        .reverse
+        .headOption
+    } else {
+      Some(0, None)
```

Comment:
What does (0, None) mean? If it means that the linage starts from 1 and we don't have any snapshot uploaded yet, we should use None to indicate that, and the return value should be `Option[(Long, String)]. (0, None) is not consistent with how we deal with checkpointID. A checkpointID None usually means it is V1 so checkpointID is not available.

Reply:
changed